src/probability/martingale/optional_sampling.lean

```diff
@@ -0,0 +1,266 @@
+/-
+Copyright (c) 2023 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne
+-/
+
+import order.succ_pred.linear_locally_finite
+import probability.martingale.basic
+
+/-!
+# Optional sampling theorem
+
+If `τ` is a bounded stopping time and `σ` is another stopping time, then the value of a martingale
+`f` at the stopping time `min τ σ` is almost everywhere equal to
+`μ[stopped_value f τ | hσ.measurable_space]`.
+
+## Main results
+
+* `stopped_value_ae_eq_condexp_of_le_const`: the value of a martingale `f` at a stopping time `τ`
+  bounded by `n` is the conditional expectation of `f n` with respect to the σ-algebra generated by
+  `τ`.
+* `stopped_value_ae_eq_condexp_of_le`: if `τ` and `σ` are two stopping times with `σ ≤ τ` and `τ` is
+  bounded, then the value of a martingale `f` at `σ` is the conditional expectation of its value at
+  `τ` with respect to the σ-algebra genrated by `σ`.
+* `stopped_value_min_ae_eq_condexp` : the optional sampling theorem: if `τ` is a bounded stopping
+  time and `σ` is another stopping time, then the value of a martingale `f` at the stopping time
+  `min τ σ` is almost everywhere equal to `μ[stopped_value f τ | hσ.measurable_space]`.
+
+-/
+
+open_locale measure_theory big_operators ennreal
+open topological_space
+
+-- TODO after the port: move to topology/instances/discrete
+@[priority 100]
+instance discrete_topology.second_countable_topology_of_countable {α : Type*} [topological_space α]
+  [hd : discrete_topology α] [countable α] :
+  second_countable_topology α :=
+@discrete_topology.second_countable_topology_of_encodable _ _ _ (encodable.of_countable _)
+
+namespace measure_theory
+
+namespace martingale
+
+variables {Ω E : Type*} {m : measurable_space Ω} {μ : measure Ω}
+  [normed_add_comm_group E] [normed_space ℝ E] [complete_space E]
+
+section first_countable_topology
+
+variables {ι : Type*} [linear_order ι] [topological_space ι] [order_topology ι]
+  [first_countable_topology ι]
+  {ℱ : filtration ι m} [sigma_finite_filtration μ ℱ] {τ σ : Ω → ι} {f : ι → Ω → E}  {i n : ι}
+
+lemma condexp_stopping_time_ae_eq_restrict_eq_const
+  [(filter.at_top : filter ι).is_countably_generated]
+  (h : martingale f ℱ μ) (hτ : is_stopping_time ℱ τ) [sigma_finite (μ.trim hτ.measurable_space_le)]
+  (hin : i ≤ n) :
+  μ[f n | hτ.measurable_space] =ᵐ[μ.restrict {x | τ x = i}] f i :=
```

Comment:
It's a little bit disturbing to have a martingale `f` and a filtration `ℱ` here, while when one jumps to the `stoppping.lean` file then `f` is the filtration there. It would probably be a good idea to rename the filtration to `ℱ` in this file also, but that's for another PR.